### PR TITLE
Add methods to judge symmetric and antisymmetric

### PR DESCRIPTION
**Pull request by 11231f**
* `bool Matrix<ELEMENT_TYPE>::symmetric()`
* `bool Matrix<ELEMENT_TYPE>::antisymmetric()`

See #48, #113.

### CODE REVIEW

File: src/include/mca/__mca_internal/single_thread_matrix_calculation.h [hunk 1]
```diff
@@ -312,8 +312,6 @@ void transposeSingleThread(const Matrix<T> &a,
 /* Check whether or not rows of matrix are symmetric
```
Comment (Kaiser-Yang) [inline src/include/mca/__mca_internal/single_thread_matrix_calculation.h:312]: Update this and make it be similar with the pic below:
image

File: src/include/mca/__mca_internal/single_thread_matrix_calculation.h [hunk 2]
```diff
@@ -329,12 +327,10 @@ bool symmetricSingleThread(const Matrix<T> &a,
 /* Check whether or not rows of matrix are antisymmetric
```
Comment (Kaiser-Yang) [inline src/include/mca/__mca_internal/single_thread_matrix_calculation.h:327]: Same requirements as the previous one.

File: src/include/mca/matrix.h [hunk 3]
```diff
@@ -569,9 +569,58 @@ void Matrix<T>::pow(const size_type &exponent, Matrix<O> &output) const {
 
 // TODO
```
Comment (Kaiser-Yang) [inline src/include/mca/matrix.h:570]: remove this `TODO`

File: test/src/multi_thread_matrix_test.cpp [hunk 4]
```diff
@@ -517,5 +517,65 @@ TEST_F(TestMatrixMultiThread, transpose) {
     ASSERT_EQ(singleOutput, multiOutput);
 }
 
+TEST_F(TestMatrixMultiThread, symmetric) {
+    a = Matrix<double>(squareShape);
+    for (size_t i = 0; i < a.rows(); i++) {
+        for (size_t j = 0; j < a.columns(); j++) { a.get(i, j) = generator() % MAX_VALUE; }
+    }
+    b                  = Matrix<double>(rectangleShape);
+    c                  = Matrix<double>(squareShape);
```
Comment (Kaiser-Yang) [inline test/src/multi_thread_matrix_test.cpp:526]: Use random number to generator a symmetric matrix.

File: test/src/multi_thread_matrix_test.cpp [hunk 5]
```diff
@@ -517,5 +517,65 @@ TEST_F(TestMatrixMultiThread, transpose) {
...
+    bool multiBool2 = b.symmetric();
+    bool multiBool3 = c.symmetric();
+    endTime         = high_resolution_clock::now();
+    executionTime   = duration_cast<std::chrono::milliseconds>(endTime - startTime).count();
+    // record multi-thread time in gtest
+    testing::Test::RecordProperty("MultiTime", executionTime);
+
+    // make sure they are equal
+    ASSERT_EQ(singleBool1, multiBool1);
+    ASSERT_EQ(singleBool2, multiBool2);
```
Comment (Kaiser-Yang) [inline test/src/multi_thread_matrix_test.cpp:546]: add those:
* `ASSERT_FALSE(singleBool2)`
* `ASSERT_TRUE(singleBool3)`

File: test/src/multi_thread_matrix_test.cpp [hunk 6]
```diff
@@ -517,5 +517,65 @@ TEST_F(TestMatrixMultiThread, transpose) {
...
+    ASSERT_EQ(singleBool3, multiBool3);
+}
+
+TEST_F(TestMatrixMultiThread, antisymmetric) {
+    a = Matrix<double>(squareShape);
+    for (size_t i = 0; i < a.rows(); i++) {
+        for (size_t j = 0; j < a.columns(); j++) { a.get(i, j) = generator() % MAX_VALUE; }
+    }
+    b                  = Matrix<double>(rectangleShape);
+    c                  = Matrix<double>(squareShape);
```
Comment (Kaiser-Yang) [inline test/src/multi_thread_matrix_test.cpp:556]: `c` is not an antisymmetrix matrix.

File: test/src/multi_thread_matrix_test.cpp [hunk 7]
```diff
@@ -517,5 +517,65 @@ TEST_F(TestMatrixMultiThread, transpose) {
...
+    bool multiBool2 = b.antisymmetric();
+    bool multiBool3 = c.antisymmetric();
+    endTime         = high_resolution_clock::now();
+    executionTime   = duration_cast<std::chrono::milliseconds>(endTime - startTime).count();
+    // record multi-thread time in gtest
+    testing::Test::RecordProperty("MultiTime", executionTime);
+
+    // make sure they are equal
+    ASSERT_EQ(singleBool1, multiBool1);
+    ASSERT_EQ(singleBool2, multiBool2);
```
Comment (Kaiser-Yang) [inline test/src/multi_thread_matrix_test.cpp:576]: Make the `c` an antisymmetric matrix and add those:
* `ASSERT_FALSE(singleBool2)`
* `ASSERT_TRUE(singleBool3)`

**Review by Kaiser-Yang — CHANGES_REQUESTED** [review]

File: test/src/multi_thread_matrix_test.cpp [hunk 8]
```diff
@@ -517,5 +517,65 @@ TEST_F(TestMatrixMultiThread, transpose) {
...
+    }
+    b                  = Matrix<double>(rectangleShape);
+    c                  = Matrix<double>(squareShape);
+    auto startTime     = high_resolution_clock::now();
+    bool singleBool1   = a.symmetric();
+    bool singleBool2   = b.symmetric();
+    bool singleBool3   = c.symmetric();
+    auto endTime       = high_resolution_clock::now();
+    auto executionTime = duration_cast<std::chrono::milliseconds>(endTime - startTime).count();
+    // get the multi-thread mode time
```
Comment (Kaiser-Yang) [inline test/src/multi_thread_matrix_test.cpp:533]: You forget to use 'init(THREAD_NUM)` to initialize the `mca`.

File: test/src/multi_thread_matrix_test.cpp [hunk 9]
```diff
@@ -517,5 +517,65 @@ TEST_F(TestMatrixMultiThread, transpose) {
...
+    }
+    b                  = Matrix<double>(rectangleShape);
+    c                  = Matrix<double>(squareShape);
+    auto startTime     = high_resolution_clock::now();
+    bool singleBool1   = a.antisymmetric();
+    bool singleBool2   = b.antisymmetric();
+    bool singleBool3   = c.antisymmetric();
+    auto endTime       = high_resolution_clock::now();
+    auto executionTime = duration_cast<std::chrono::milliseconds>(endTime - startTime).count();
+    // get the multi-thread mode time
```
Comment (Kaiser-Yang) [inline test/src/multi_thread_matrix_test.cpp:563]: You forget to use 'init(THREAD_NUM)` to initialize the `mca`.

**Review by Kaiser-Yang — CHANGES_REQUESTED** [review]

File: src/include/mca/matrix.h [hunk 10]
```diff
@@ -567,11 +567,59 @@ void Matrix<T>::pow(const size_type &exponent, Matrix<O> &output) const {
     }
 }
 
-// TODO
 template <class T>
-bool Matrix<T>::symmetric() const noexcept {}
+bool Matrix<T>::symmetric() const noexcept {
```
Comment (Kaiser-Yang) [inline src/include/mca/matrix.h:571]: remove the `noexcept` here and at the declarations.

File: src/include/mca/matrix.h [hunk 11]
```diff
@@ -567,11 +567,59 @@ void Matrix<T>::pow(const size_type &exponent, Matrix<O> &output) const {
...
+    bool result = symmetricSingleThread(
+        *this, (res.second - 1) * res.first, (size() - (res.second - 1) * res.first), epsilon());
+
+    // make sure all the sub threads are finished
+    for (auto &item : returnValue) { result &= item.get(); }
+    return result;
+}
 template <class T>
-bool Matrix<T>::antisymmetric() const noexcept {}
+bool Matrix<T>::antisymmetric() const noexcept {
```
Comment (Kaiser-Yang) [inline src/include/mca/matrix.h:597]: remove the `noexcept`

File: src/include/mca/matrix.h [hunk 12]
```diff
@@ -567,11 +567,59 @@ void Matrix<T>::pow(const size_type &exponent, Matrix<O> &output) const {
...
+    // let main thread calculate took
+    bool result = antisymmetricSingleThread(
+        *this, (res.second - 1) * res.first, (size() - (res.second - 1) * res.first), epsilon());
+
+    // make sure all the sub threads are finished
+    for (auto &item : returnValue) { result &= item.get(); }
+    return result;
+}
 }  // namespace mca
+   // namespace mca
```
Comment (Kaiser-Yang) [inline src/include/mca/matrix.h:623]: remove this line

File: test/src/multi_thread_matrix_test.cpp [hunk 13]
```diff
@@ -517,5 +517,96 @@ TEST_F(TestMatrixMultiThread, transpose) {
...
 }
 
+TEST_F(TestMatrixMultiThread, symmetric) {
+    a = Matrix<double>(squareShape);
+    for (size_t i = 0; i < a.rows(); i++) {
+        for (size_t j = 0; j < a.columns(); j++) { a.get(i, j) = generator() % MAX_VALUE; }
+    }
+    b = Matrix<double>(rectangleShape);
+    c = Matrix<double>(squareShape);
+    for (size_t i = 0; i < a.rows(); i++) {
```
Comment (Kaiser-Yang) [inline test/src/multi_thread_matrix_test.cpp:527]: update `a.rows` with `c.rows` and update `a.columns` with `c.columns`

File: test/src/multi_thread_matrix_test.cpp [hunk 14]
```diff
@@ -517,5 +517,96 @@ TEST_F(TestMatrixMultiThread, transpose) {
...
+}
+
+TEST_F(TestMatrixMultiThread, antisymmetric) {
+    a = Matrix<double>(squareShape);
+    for (size_t i = 0; i < a.rows(); i++) {
+        for (size_t j = 0; j < a.columns(); j++) { a.get(i, j) = generator() % MAX_VALUE; }
+    }
+    b = Matrix<double>(rectangleShape);
+    c = Matrix<double>(squareShape);
+    for (size_t i = 0; i < a.rows(); i++) {
```
Comment (Kaiser-Yang) [inline test/src/multi_thread_matrix_test.cpp:572]: update `a` with `c`.

File: test/src/multi_thread_matrix_test.cpp [hunk 15]
```diff
@@ -517,5 +517,96 @@ TEST_F(TestMatrixMultiThread, transpose) {
...
+TEST_F(TestMatrixMultiThread, symmetric) {
+    a = Matrix<double>(squareShape);
+    for (size_t i = 0; i < a.rows(); i++) {
+        for (size_t j = 0; j < a.columns(); j++) { a.get(i, j) = generator() % MAX_VALUE; }
+    }
+    b = Matrix<double>(rectangleShape);
+    c = Matrix<double>(squareShape);
+    for (size_t i = 0; i < a.rows(); i++) {
+        for (size_t j = 0; j < a.columns(); j++) {
+            if (i < j) (c.get(i, j) = c.get(j, i) = generator() % MAX_VALUE);
```
Comment (Kaiser-Yang) [inline test/src/multi_thread_matrix_test.cpp:529]: add `{}`.

File: test/src/multi_thread_matrix_test.cpp [hunk 16]
```diff
@@ -517,5 +517,96 @@ TEST_F(TestMatrixMultiThread, transpose) {
...
+TEST_F(TestMatrixMultiThread, antisymmetric) {
+    a = Matrix<double>(squareShape);
+    for (size_t i = 0; i < a.rows(); i++) {
+        for (size_t j = 0; j < a.columns(); j++) { a.get(i, j) = generator() % MAX_VALUE; }
+    }
+    b = Matrix<double>(rectangleShape);
+    c = Matrix<double>(squareShape);
+    for (size_t i = 0; i < a.rows(); i++) {
+        for (size_t j = 0; j < a.columns(); j++) {
+            if (i < j) (c.get(i, j) = generator() % MAX_VALUE);
```
Comment (Kaiser-Yang) [inline test/src/multi_thread_matrix_test.cpp:574]: add `{}`.

**Review by Kaiser-Yang — CHANGES_REQUESTED** [review]

**Review by Quentin9922 — APPROVED** [review]

**Review by Kaiser-Yang — APPROVED** [review]

**Pull request merged**